handler.c

```diff
@@ -1,562 +1,546 @@
+
+  if (!plist) {
+    fprintf(stderr, "%s: Invalid plist\n", __FUNCTION__);
+    return (1);
+  }
+
+  if ((dharray = CFDictionaryGetValue(plist, DUTI_KEY_SETTINGS)) == NULL) {
+    fprintf(stderr, "%s is missing the settings array\n", spath);
+    CFRelease(plist);
+    return (1);
+  }
+
+  count = CFArrayGetCount(dharray);
+  for (index = 0; index < count; index++) {
+    dhentry = CFArrayGetValueAtIndex(dharray, index);
+
+    if ((bid = CFDictionaryGetValue(dhentry, DUTI_KEY_BUNDLEID)) == NULL) {
+      fprintf(stderr, "Entry %d missing bundle ID\n", (int)index);
+      rc = 1;
+      continue;
+    }
+    if ((url_scheme = CFDictionaryGetValue(dhentry, DUTI_KEY_URLSCHEME)) !=
+        NULL) {
+      htype = DUTI_TYPE_URL_HANDLER;
+    } else if ((uti = CFDictionaryGetValue(dhentry, DUTI_KEY_UTI)) != NULL) {
+      if ((role = CFDictionaryGetValue(dhentry, DUTI_KEY_ROLE)) == NULL) {
+        fprintf(stderr, "Entry %d missing role\n", (int)index);
+        rc = 1;
+        continue;
+      }
+      htype = DUTI_TYPE_UTI_HANDLER;
+    } else {
+      fprintf(stderr, "Entry %d invalid\n", (int)index);
+      rc = 1;
+      continue;
+    }
+
+    if (htype == DUTI_TYPE_UTI_HANDLER) {
+      if (cf2c(bid, handler, sizeof(handler)) != 0) {
+        rc = 1;
+        continue;
+      }
+      if (cf2c(uti, type, sizeof(type)) != 0) {
+        rc = 1;
+        continue;
+      }
+      if (cf2c(role, crole, sizeof(crole)) != 0) {
+        rc = 1;
+        continue;
+      }
+      if (duti_handler_set(handler, type, crole) != 0) {
+        rc = 1;
+      }
+    } else if (htype == DUTI_TYPE_URL_HANDLER) {
+      if (cf2c(bid, handler, sizeof(handler)) != 0) {
+        rc = 1;
+        continue;
+      }
+      if (cf2c(url_scheme, type, sizeof(type)) != 0) {
+        rc = 1;
+        continue;
+      }
+      if (duti_handler_set(handler, type, NULL) != 0) {
+        rc = 1;
+      }
+    }
+  }
+
+  if (plist != NULL) {
+    CFRelease(plist);
+  }
+
+  return (rc);
 }
 
-    int
-set_uti_handler( CFStringRef bid, CFStringRef type, LSRolesMask mask )
-{
-    OSStatus		rc;
-    char		cb[ MAXPATHLEN ];
-    char		ct[ MAXPATHLEN ];
+int dirsethandler(char *dirpath) {
+  DIR *d = NULL;
+  struct dirent *de;
+  struct ll *head = NULL;
+  struct ll *cur, *tmp;
+  struct stat st;
+  char path[MAXPATHLEN];
+  char *p;
+  int rc = 0;
+  int (*dhandler_f)(char *);
 
-    /* get C strings to use in output. CFShow is inadequate. */
-    if ( cf2c( bid, cb, sizeof( cb )) != 0 ) {
-	/* something reasonable on error */
-	strlcpy( cb, "bundle_id", sizeof( cb ));
-    }
-    if ( cf2c( type, ct, sizeof( ct )) != 0 ) {
-	strlcpy( ct, "UTI", sizeof( ct ));
-    }
+  if ((d = opendir(dirpath)) == NULL) {
+    fprintf(stderr, "opendir %s: %s\n", dirpath, strerror(errno));
+    exit(2);
+  }
 
-    
-    if ( !duti_is_conformant_uti( type )) {
-	fprintf( stderr, "%s does not conform to any UTI hierarchy\n", ct );
-	return( 1 );
+  while ((de = readdir(d)) != NULL) {
```

Comment:
To what end? Are we concerned about holding this storage for the length of the function?